### ISSUE TITLE
predict 1s audio

### ISSUE THREAD

**Issue by LucySha** [MaigoAkisame/cmu-thesis#3]
@MaigoAkisame Thanks for your work. I have 2 questions below, hope you can help me figure out.
1. input 400 frames, the params: win_length and hop length (maybe 64ms per frame with 25ms hop_length)
2. the input size fixed to 400 frames, if the input audio lasts 1s, can the net work?  or if I have a audio lasts 10s, I want to predict the label every 1s, does it work?

**Comment by MaigoAkisame** [comment]
1. Yes, win_length is 64 ms and hop_length is 25 ms.
2. Yes, you can feed in an audio of arbitrary length, and the network should work.
3. The network always produces labels every 0.1 second. If you want labels every 1 second, you'll need to aggregate the labels in groups of 10. I recommend using the linear softmax pooling function for aggregation.

**Comment by LucySha** [comment]
@MaigoAkisame thanks for your reply, but I'm confused with the input and output, the net input need 400frames (10s), does that mean I have to provide 10s audio as input to get the label, If the audio was 1s(40 frames), how to ensure we have 400 frames features as input, zero padding ?

**Comment by MaigoAkisame** [comment]
The input doesn't have to be exactly 10 seconds.

If you get an error feeding in an audio of 1 second, can you post the error message?

Or, you can try zero-padding it to 10 seconds.

**Comment by MaigoAkisame** [comment]
I haven't kept a pre-trained model, so I'm afraid you need to train one using the code and data provided in this repo.

**Comment by LucySha** [comment]
@MaigoAkisame Before training, you did normalization to the features(data['feat'-mu]/sigma), after predicting , you did (global_prob.clamp_(min=1e-7, max=1-1e-7)) Is it necessary? how did you get 'mu' and 'sigma'

**Comment by LucySha** [comment]
And I was also confused with the params , according to 'cmu-thesis', chapter3,  the frames was 30ms, hop was 6.25ms, and the fiterbank of 40 triangle filters spans a frequency range from 0 Hz to 5.8kHz, it means that the input batch*400*64, the '64' =40(0 Hz to 5.8kHz) + 24(5.8kHz to 8kHz), am I right?

**Comment by LucySha** [comment]
and another question ,how to map the model output to event types, for example, the output of the audioset is 1 * 527 vector, how to map the vector(probability) to event types?

**Comment by MaigoAkisame** [comment]
@LucySha 

1. For feature normalization, the mu and sigma are computed from all the training data.

2. The clamping can avoid probabilities underflowing to zero and causing all kinds of nasty problems (such as NaNs).

3. The mapping between event IDs and event names follows this file (found on the "download" page of the AudioSet website):
[](http://storage.googleapis.com/us_audioset/youtube_corpus/v1/csv/class_labels_indices.csv)

**Comment by LucySha** [comment]
@MaigoAkisame Thanks a lot! I'll train the audioset network to help me understand the task better.

**Comment by LucySha** [comment]
@MaigoAkisame  I trained the 'dcase' model andt it works, but when I trained the 'audioset',  'util_in.py' line 41 went error due to the large number of N_CLASSES,  error msg:'OSERROR, too many open files', Can I reduce the number of N_CLASSES(it meant 527 classes)

**Comment by LucySha** [comment]
@MaigoAkisame I have trained the audioset net, thanks again! but I have another question about the audioset feature you provided, for example 'GAS_train_unbalanced_class233_part0014.mat', class233 stands for the audio labels contain class233, and part0014 stands for the index, different classes have different numbers, am I right?  If an audio has both class233 and class234 labels, Does 'GAS_train_unbalanced_class233' and 'GAS_train_unbalanced_class234' both have the same audio?

**Comment by MaigoAkisame** [comment]
@LucySha Yes, an audio will appear in multiple archives if it carries multiple labels.
